src/lib.rs

```diff
@@ -146,16 +152,31 @@ pub struct ValidatedMessage {
     /// URI of the schema validating this message.
     ///
     /// E.g. `https://hedwig.domain.xyz/schemas#/schemas/user.created/1.0`
-    schema: &'static str,
+    schema: Cow<'static, str>,
     /// Custom message headers.
     ///
     /// This may be used to track request_id, for example.
     headers: Headers,
     /// The encoded message data.
-    data: Vec<u8>,
+    data: Bytes,
 }
 
 impl ValidatedMessage {
+    /// Create a new validated message
+    pub fn new<S, D>(id: Uuid, timestamp: SystemTime, schema: S, headers: Headers, data: D) -> Self
```

Comment:
I don't think we should have a public method like this. I understand that you did it to allow for 3rd party implementations of `Consumer` to happen, but on the other hand this allows for the publisher side to skip validation entirely.

Maybe it would make sense to have two separate types for the encoded message for the publisher and the subscriber side?

Reply:
Having ValidatedMessage private-constructed also means that users can't implement their own validators. That seems like a particularly harsh restriction -- i would consider it worse than the risk of a user manually constructing messages